### PR TITLE
enhancement(vrl): add fuzz testing for the VRL VM

### PR DESCRIPTION
**Pull request by StephenWakely**
This introduces a first version of fuzz testing for the VRL VM.

The program generated to test against is incredibly basic, essentially a single line expression returning a string, integer or boolean literal. It's aim is to just set up the initial framework and test the concept.

You can run the test by:

```sh
> cd lib/vrl/compiler
> cargo +nightly fuzz run vm_differential --release --debug-assertions -s address --jobs $(nproc) -- -max_len=65536
```

### CODE REVIEW

File: rust-toolchain.toml [hunk 1]
```diff
@@ -1,3 +0,0 @@
-[toolchain]
```
Comment (StephenWakely) [inline rust-toolchain.toml:1]: This is a problem that needs resolving. It seems `cargo fuzz` requires running on nightly, but our rust-toolchain forces you to run against `1.57` - which stops the fuzz test from runnning.

Would welcome solutions to this!

Comment (jszwedko) [reply rust-toolchain.toml:1]: 🤔 Is it not possible to just pass `+nightly` when running `cargo` to build the fuzz tests? We could add this as a `make` target.

Comment (StephenWakely) [reply rust-toolchain.toml:1]: Sweet! Yes that works!

File: lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#![no_main]
+use libfuzzer_sys::fuzz_target;
+use std::collections::BTreeMap;
+use vrl_compiler as vrl;
+
+fuzz_target!(|data: Vec<parser::Program>| {
+    for expr in data {
+        // Compile the VRL.
+        let program = vrl::compile(expr, &vec![]).unwrap();
```
Comment (pablosichert) [inline lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs:9]: Does this take into account that there's probably a lot of ASTs that will not pass static analysis? Or are the programs passed into the closure already pre-validated?

File: lib/vrl/compiler/fuzz/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+target
+corpus
```
Comment (blt) [inline lib/vrl/compiler/fuzz/.gitignore:2]: The corpus would be valuable to record so as to prime other users' runs.

File: lib/vrl/compiler/fuzz/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "vrl-compiler-fuzz"
+version = "0.1.0"
+authors = ["Vector Contributors <vector@timber.io>"]
```
[the remark this answers is not in the document]
Comment (StephenWakely) [reply lib/vrl/compiler/fuzz/Cargo.toml:4]: Ah, good spot. There's a bunch of VRL crates that need updating too.

File: lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs (hunk 2 above)
Comment (StephenWakely) [reply lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs:9]: That's a good question. Not at the minute. With this an invalid AST would cause a panic which would indicate a test failure. I need to work out how to set it up so invalid ASTs are just ignored..

File: lib/vrl/compiler/fuzz/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply lib/vrl/compiler/fuzz/Cargo.toml:4]: Was mentioned to Jean recently as well, should be a reasonably quick find an replace PR

File: lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
+#![no_main]
```
Comment (blt) [inline lib/vrl/compiler/fuzz/fuzz_targets/fuzz_target_1.rs:1]: You might give this file a more descriptive name than `fuzz_target_1`.

File: lib/vrl/parser/src/arbitrary.rs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        FunctionCall, Group, Ident, IfStatement, Node, Op, Opcode, Predicate, Query, QueryTarget,
+        RootExpr,
+    },
+    Literal, Program,
+};
+use arbitrary::{Arbitrary, Unstructured};
+use diagnostic::Span;
+use lookup::LookupBuf;
+
+const DEPTH: isize = 4;
```
Comment (blt) [inline lib/vrl/parser/src/arbitrary.rs:20]: Fuzzing turn up anything interesting if you increase this, or does increasing this blow up the run?

**Review by blt — APPROVED** [review]
Very nice. I had one small comment about a file name -- no blocker there, stray thought -- and a question.

File: lib/vrl/parser/src/arbitrary.rs (hunk 6 above)
Comment (StephenWakely) [reply lib/vrl/parser/src/arbitrary.rs:20]: The bigger the number, it seems the odds of generating a valid VRL program rapidly approach zero - so no tests get run.

Comment (blt) [reply lib/vrl/parser/src/arbitrary.rs:20]: Heard, makes sense.

**Comment by StephenWakely** [comment]
> I'd appreciate however if the fixes in response to the fuzz testing in here would be split out into separate PRs and the branch be rebased on top of master rather than the old VM branch. It's a bit hard to follow and reason about otherwise.

I have rebased it now. I haven't split it out into separate PRs however as I feel (hope) that it is simple enough to follow what is going on now. Basically anything under `fuzz` and `arbitrary` is specific to the fuzz testing, everything else is fixes to bugs found during the testing..

**Comment by jszwedko** [comment]
Merged this since Pablo looked at it. The soak test seems to be a flake given the baseline looks erratic and VRL isn't involved in that soak.
